sdk/src/main/scala/com/horizen/SidechainWallet.scala

```diff
@@ -309,6 +312,58 @@ class SidechainWallet private[horizen] (seed: Array[Byte],
   def getCswData(withdrawalEpochNumber: Int): Seq[CswData] = {
     cswDataStorage.getCswData(withdrawalEpochNumber)
   }
+
+  // Check that all wallet storages are consistent and in case forging box info storage is not, then try a rollback for it.
+  // Return the state and common version or throw an exception if some unrecoverable misalignment has been detected
+  def ensureStorageConsistencyAfterRestore: Try[(SidechainWallet, ByteArrayWrapper)] = {
+    // TODO csw capability will be optional, related storage might even be empty
+    val version = walletBoxStorage.lastVersionId
+    if (
+      // check these storages first, they are updated always together
+        version == walletTransactionStorage.lastVersionId &&
+          version == cswDataStorage.lastVersionId
+    ) {
+      // check forger box info storage, which is updated before state in case of epoch switch
+      if (version == forgingBoxesInfoStorage.lastVersionId) {
+        log.debug("All wallet storage versions are consistent")
+        Success(this, version.get)
+      } else {
+        val versionSome = version.get
+        // the version should be the previous at most
+        val maxNumberOfVersionToRetrieve = 2
+        val rollbackList = forgingBoxesInfoStorage.rollbackVersions(maxNumberOfVersionToRetrieve)
+        if (rollbackList.last == versionSome) {
+          if (forgingBoxesInfoStorage.size == maxNumberOfVersionToRetrieve) {
+            // this is ok, we have just the genesis block whose modId is the very first version, and the second
+            // is the non-rollback version used when updating the ForgingStakeMerklePathInfo at the startup
+            log.debug("All wallet storage versions are consistent")
+            Success(this, version.get)
+          } else {
+            // it can be the case of process stopped when we had not yet updated the state and we are switching epoch
+            log.warn(s"Wallet forger box storage versions is NOT consistent, trying to rollback")
+            val rbVersion = forgingBoxesInfoStorage.rollback(versionSome) match {
+              case (Success(_)) => Some(bytesToVersion(versionSome.data()))
+              case Failure(exception) => {
+                None
+              }
+            }
+            if (rbVersion.isEmpty) {
+              // unrecoverable
+              log.error("Could not recover wallet forging storages")
+              throw new RuntimeException("Could not rollback wallet forging box info storages")
+            }
+            Success(this, version.get)
+          }
+        } else {
+          log.error("Forging boxes info storage does not have the expected version in the rollback list")
+          throw new RuntimeException("Forging boxes info storage does not have the expected version in the rollback list")
+        }
+      }
+    } else {
```

Comment:
The reason is because when we check wallet storage consistency, we already ensured that state and history are consistent, and since wallet is update in between state and history (state --> wallet --> history) we can not have any wallet inconsistency and therefore should throw an exception.